exporters/otlp/src/otlp_grpc_client.cc

```diff
@@ -140,31 +365,216 @@ OtlpGrpcClient::MakeLogsServiceStub(const OtlpGrpcClientOptions &options)
+    std::unique_ptr<google::protobuf::Arena> &&arena,
+    proto::collector::metrics::v1::ExportMetricsServiceRequest &&request,
+    std::function<bool(opentelemetry::sdk::common::ExportResult,
+                       std::unique_ptr<google::protobuf::Arena> &&,
+                       const proto::collector::metrics::v1::ExportMetricsServiceRequest &,
+                       proto::collector::metrics::v1::ExportMetricsServiceResponse *)>
+        &&result_callback) noexcept
+{
+  auto metrics_count = request.resource_metrics_size();
+  if (is_shutdown_)
+  {
+    OTEL_INTERNAL_LOG_ERROR("[OTLP GRPC Client] ERROR: Export "
+                            << metrics_count << " metric(s) failed, exporter is shutdown");
+    if (result_callback)
+    {
+      result_callback(opentelemetry::sdk::common::ExportResult::kFailure, std::move(arena), request,
+                      nullptr);
+    }
+    return opentelemetry::sdk::common::ExportResult::kFailure;
+  }
+
+  std::shared_ptr<OtlpGrpcClientAsyncData> async_data = MutableAsyncData(options);
+  return InternalDelegateAsyncExport(async_data, stub, std::move(context), std::move(arena),
+                                     std::move(request), std::move(result_callback), metrics_count,
+                                     "metric(s)");
+}
+
+sdk::common::ExportResult OtlpGrpcClient::DelegateAsyncExport(
+    const OtlpGrpcClientOptions &options,
+    proto::collector::logs::v1::LogsService::StubInterface *stub,
+    std::unique_ptr<grpc::ClientContext> &&context,
+    std::unique_ptr<google::protobuf::Arena> &&arena,
+    proto::collector::logs::v1::ExportLogsServiceRequest &&request,
+    std::function<bool(opentelemetry::sdk::common::ExportResult,
+                       std::unique_ptr<google::protobuf::Arena> &&,
+                       const proto::collector::logs::v1::ExportLogsServiceRequest &,
+                       proto::collector::logs::v1::ExportLogsServiceResponse *)>
+        &&result_callback) noexcept
+{
+  auto logs_count = request.resource_logs_size();
+  if (is_shutdown_)
+  {
+    OTEL_INTERNAL_LOG_ERROR("[OTLP GRPC Client] ERROR: Export "
+                            << logs_count << " log(s) failed, exporter is shutdown");
+    if (result_callback)
+    {
+      result_callback(opentelemetry::sdk::common::ExportResult::kFailure, std::move(arena), request,
+                      nullptr);
+    }
+    return opentelemetry::sdk::common::ExportResult::kFailure;
+  }
+
+  std::shared_ptr<OtlpGrpcClientAsyncData> async_data = MutableAsyncData(options);
+  return InternalDelegateAsyncExport(async_data, stub, std::move(context), std::move(arena),
+                                     std::move(request), std::move(result_callback), logs_count,
+                                     "log(s)");
+}
+
+std::shared_ptr<OtlpGrpcClientAsyncData> OtlpGrpcClient::MutableAsyncData(
+    const OtlpGrpcClientOptions &options)
+{
+  if (!async_data_)
```

Comment:
In my understanding, `MutableAsyncData()` can be called concurrently by several threads.

When 2 threads call this for the first time, when `async_data_` is still nullptr, is there a race condition here, with two threads invoking std::make_shared() and competing to assign `async_data_` ?

What if thread 1 returns async_data_ allocated first, and then thread 2 assign a second async_data_, the first shared pointer returned is no longer referenced internally, and could cause issues during flush / shutdown.

Use swap instead of an assignment, so at least the caller don't see a partially initialized async data ?

Reply:
According to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#exportbatch and in my understanding. `Export()` will never be called concurrently, and `MutableAsyncData()` will only be called in `Export()` ?

Reply:
Ok, thanks for the clarification. Resolved.